File: Frontend/src/app/services/shape-factory.ts
import { Injectable } from '@angular/core';
import { Circle } from '../models/circle';
import { Rectangle } from '../models/rectangle';
import { Line } from '../models/line';
import { Triangle } from '../models/triangle';
import { Freehand } from '../models/freehand';
import { Shape } from '../models/shape';
import { Ellipse } from '../models/ellipse';
import {Square} from '../models/square';
@Injectable({
  providedIn: 'root'
})
export class ShapeFactoryService {

  createShape(
    type: string,
    x1: number,
    y1: number,
    x2: number,
    y2: number,
    properties: any
  ): Shape {

    switch (type) {
      case 'circle':
        return new Circle(x1, y1, x2, y2, properties);

      case 'rectangle':
        return new Rectangle(x1, y1, x2, y2, properties);

      case 'square':
        return new Square(x1, y1, x2, y2, properties);

      case 'line':
        return new Line(x1, y1, x2, y2, properties);

      case 'freehand':
        return new Freehand(x1, y1, x2, y2, properties);

      default:
        throw new Error(`Unknown shape type: ${type}`);
    }
  }
}

File: Backend/demo/src/main/java/com/example/paintapp/dtos/ShapeFactory.java
package com.example.paintapp.dtos;

import com.example.paintapp.services.ShapeDTO;
import com.example.paintapp.shapes.*;
import com.example.paintapp.shapes.base.Shape;

public class ShapeFactory {

    public static Shape  createShape(ShapeDTO dto) {

        if (dto == null || dto.getType() == null) {
            throw new IllegalArgumentException("Invalid ShapeDTO");
        }

        Shape shape;

        switch (dto.getType().toLowerCase()) {

            case "circle":
                shape = new Circle();
                break;

            case "ellipse":
                shape = new Ellipse();
                break;

            case "freehand":
                shape = new Freehand();
                break;

            case "line":
                shape = new Line();
                break;

            case "rectangle":
                shape = new Rectangle();
                break;

            case "square":
                shape = new Square();
                break;

            case "triangle":
                shape = new Triangle();
                break;

            default:
                throw new IllegalArgumentException("Unknown shape type: " + dto.getType());
        }

        // Set shared fields
        shape.setType(dto.getType());
        shape.setX1(dto.getX1());
        shape.setY1(dto.getY1());
        shape.setX2(dto.getX2());
        shape.setY2(dto.getY2());

        // Set all extra properties (radius, width, height, etc.)
        if (dto.getProperties() != null) {
            shape.setProperties(dto.getProperties());
        }

        return shape;
    }
}
